### DISCUSSION TITLE
Getting Markdown files layout to match the Project Tree

### DISCUSSION THREAD

**Discussion by krvoje** [vkbo/novelWriter#839] [Ideas]
Basically, to have the novel laid out in a directory structure that mimics the Project Tree view, instead of the content dir with hashes (or as an additional export option).

**Comment by vkbo** [comment]
You mean in the project's contents folder on your hard drive? There's a previous discussion about it in the version control issue, [here](https://github.com/vkbo/novelWriter/issues/383#issuecomment-712548532).

The short answer is that you are not supposed to touch those files. If you need to find a specific file in there for some reason or another, there's a `ToC.txt` file that tells you which file in the project tree is saved where.

There are very specific reasons why the file structure is the way it is, and I have no plans to change it. Using the internal file reference in the project as the file name makes the whole solution a lot more stable and less error prone. It also makes it significantly easier to implement the versioning features I'm planning to add.

However, I have considered an export feature that could do what you request on a one-off basis. It would make it easier to extract your work as it is structured without having to make a build.

**Reply by krvoje** [reply]
Sorry, I've missed the discussion there. Yeah, this is what I meant. This would make it straightforward to implement some custom scripting, markdown processing, or use version control directly on the text. An export would be useful for this for sure.

**Reply by vkbo** [reply]
No worries. That discussion wasn't easy to find. It took me quite a while to find it again myself.

The ToC.txt file is there to make some custom scripting possible. It should be relatively simple to parse. You can also put the entire project into a JSON file from the Build tool, which makes it easy to process with for instance a Python script.

I keep all my writing projects in git repositories with the way things are currently stored, so I don't really see the issue there. I am also working on a way to save multiple versions of files in novelWriter itself. I initially wanted to integrate git for this, but it might be too technical for the average user.

**Reply by krvoje** [reply]
Thanks, I'll check those options out. The discussion in the issue you've linked to explains it well. I was looking at this from the perspective of having a collection of Markdown files for my writing, but this is a different use case. It's just my dev mindset of distrusting an editor app to take care of everything :)

I agree on git, the terminology is way too counter-intuitive for a non-technical user

**Reply by vkbo** [reply]
Yeah, you're not the first to see it as an editor of multiple markdown files, but it really isn't.

It's why I've considered adding a file format that stores the entire project in a single file. But I haven't found a method I like yet. I'm considering making it as simple as just zip the project folder when closing the project and keep it unzipped in temp when open.

**Reply by peter88213** [reply]
> It's why I've considered adding a file format that stores the entire project in a single file. But I haven't found a method I like yet. I'm considering making it as simple as just zip the project folder when closing the project and keep it unzipped in temp when open.

Why not just put the whole content in CDATA sections in the .nwx file?

**Reply by vkbo** [reply]
I wouldn't save it as CDATA, but I have considered it, yes.

**Reply by peter88213** [reply]
I'm glad to hear that, now that I've managed to read and write File Format version 1.3 with my yWriter converter. 
By the way, it would be helpful if novelWriter would create its index automatically in case it does not find an index file.

**Reply by vkbo** [reply]
Yeah, it's a good idea. There are few cases where the index is missing though. It is (re)built automatically when the index is broken, where "broken" mostly means that the format has changed, and when a project is first made. Did you encounter any cases where there was no index?

Also, the JSON file is just a dump of the index memory between sessions. It doesn't have a "file format" per say, and it has changed many times as I've added features, or restructured things. What data do you extract from the index?

If I would move to a single file format, the index file would be dropped entirely and just rebuilt on load as all the text has to be loaded anyway.

**Reply by vkbo** [reply]
By the way, feel free to start a new discussion topic if you want to talk file formats in general. I find your input very useful, but perhaps they are a bit off topic in these other threads, and I'm sure the authors of the topic get a bunch of email notifications. :)

**Reply by peter88213** [reply]
> Did you encounter any cases where there was no index?

This happens when converting a yWriter project to novelWiter. My converter script generates only the required files, i.e. 'nwProject.nwx' and the .nwd files. Such projects can be opened in novelWriter, but the index must then be created manually before the outline becomes visible. After that also a window has to be clicked away (could be omitted by me in case of success). This slows down my manual test runs a bit.

> By the way, feel free to start a new discussion topic if you want to talk file formats in general.

About file formats I could say a thing or two, but I don't want to dispute a fundamental discussion here in any case. I probably can't bring up anything here that you haven't already thought through plenty.
I think I will open a new thread to introduce my file converter once it is ready, but that may take a while.

**Reply by vkbo** [reply]
> > Did you encounter any cases where there was no index?
> 
> This happens when converting a yWriter project to novelWiter. My converter script generates only the required files, i.e. 'nwProject.nwx' and the .nwd files. Such projects can be opened in novelWriter, but the index must then be created manually before the outline becomes visible. After that also a window has to be clicked away (could be omitted by me in case of success). This slows down my manual test runs a bit.

I see, well I can certainly add a check that re-builds the index quietly if it's empty. It may be useful in other cases too.

> > By the way, feel free to start a new discussion topic if you want to talk file formats in general.
> 
> About file formats I could say a thing or two, but I don't want to dispute a fundamental discussion here in any case. I probably can't bring up anything here that you haven't already thought through plenty. I think I will open a new thread to introduce my file converter once it is ready, but that may take a while.

Sure, sounds good.

**Comment by mlo-3** [comment]
> However, I have considered an export feature that could do what you request on a one-off basis. It would make it easier to extract your work as it is structured without having to make a build.

Hi, I was looking to same thing, an export option would be nice indeed. This would be handy in order to migrate the work project to obdsidian for exemple. And also to have a sort of backup with readable file names. Just in case ;)

By the way, thanks a lot for novelwriter ! It rocks !

**Reply by vkbo** [reply]
I'm glad you find novelWriter useful :)

I absolutely see a use case for this. Not least because it makes it possible to export a project into a format that is human readable and independent of any software tool. I've added a new feature ticket for it at #1108.
